### PR TITLE
Fy/unified model path

### PR DESCRIPTION
**Pull request by future-xy**
## Description

This PR cleans up ServerlessLLM Store design by provide dedicated interface for PyTorch and Transformers. ServerlessLLM Store now uses `model_path` instead of `model_name` as the global ID for a model checkpoint. By providing a clean API, ServerlessLLM now supports storage-aware model loading scheduling for vLLM backend.

## Changes

- Move `save_model` and `load_model` to `serverless_llm_store.transformers`
- Move 'save_dict` and `load_dict` to `serverless_llm_store.torch`
- Use `serverless_llm_store.torch.save_dict/load_dict` in vLLM
- Support and require model registration for vLLM backend

### CODE REVIEW

**Comment by SiyangShao** [comment]
It seems that the lack of GPU causes the failed tests. This PR looks good to me, but I think we could merge it after we merge the GPU workflow PR so that the tests can be run successfully.

**Comment by future-xy** [comment]
> It seems that the lack of GPU causes the failed tests. This PR looks good to me, but I think we could merge it after we merge the GPU workflow PR so that the tests can be run successfully.

`serverless_llm/store/tests/python/test_save_vllm_model.py` still doesn't work. It reports a weird bug as following:

```
INFO 09-14 17:59:00 config.py:624] Defaulting to use mp for distributed inference
INFO 09-14 17:59:00 llm_engine.py:161] Initializing an LLM engine (v0.5.0.post1) with config: model='./test_models/models--facebook--opt-1.3b/snapshots/3f5c25d0bc631cb57ac65913f76e22c2dfb61d62', speculative_config=None, tokenizer='./test_models/models--facebook--opt-1.3b/snapshots/3f5c25d0bc631cb57ac65913f76e22c2dfb61d62', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=1, download_dir='./test_models/models--facebook--opt-1.3b/snapshots/3f5c25d0bc631cb57ac65913f76e22c2dfb61d62', load_format=LoadFormat.AUTO, tensor_parallel_size=2, disable_custom_all_reduce=False, quantization=None, enforce_eager=True, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), seed=0, served_model_name=./test_models/models--facebook--opt-1.3b/snapshots/3f5c25d0bc631cb57ac65913f76e22c2dfb61d62)
(VllmWorkerProcess pid=2304287) INFO 09-14 17:59:02 multiproc_worker_utils.py:215] Worker ready; awaiting tasks
INFO 09-14 17:59:03 utils.py:637] Found nccl from library libnccl.so.2
(VllmWorkerProcess pid=2304287) INFO 09-14 17:59:03 utils.py:637] Found nccl from library libnccl.so.2
INFO 09-14 17:59:03 pynccl.py:63] vLLM is using nccl==2.20.5
(VllmWorkerProcess pid=2304287) INFO 09-14 17:59:03 pynccl.py:63] vLLM is using nccl==2.20.5
(VllmWorkerProcess pid=2304287) INFO 09-14 17:59:03 custom_all_reduce_utils.py:179] reading GPU P2P access cache from /home/fy/.config/vllm/gpu_p2p_access_cache_for_0,1.json
INFO 09-14 17:59:03 custom_all_reduce_utils.py:179] reading GPU P2P access cache from /home/fy/.config/vllm/gpu_p2p_access_cache_for_0,1.json
(VllmWorkerProcess pid=2304287) INFO 09-14 17:59:07 model_runner.py:160] Loading model weights took 1.2296 GB
INFO 09-14 17:59:07 model_runner.py:160] Loading model weights took 1.2296 GB
INFO 09-14 17:59:08 llm_engine.py:317] Overriding num_gpu_blocks=13024 with num_gpu_blocks_override=1
INFO 09-14 17:59:08 distributed_gpu_executor.py:56] # GPU blocks: 1, # CPU blocks: 2730
Saving tensors: 100% [======================================================================] 100 %
Saving tensors: 100% [======================================================================] 100 %
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ Captured stderr call -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Fetching 7 files: 100%|██████████| 7/7 [00:24<00:00,  3.44s/it]
2024-09-14 17:58:59,740 INFO worker.py:1724 -- Started a local Ray instance.
Traceback (most recent call last):
  File "/home/fy/.conda/envs/sllm-worker/lib/python3.10/multiprocessing/resource_tracker.py", line 209, in main
    cache[rtype].remove(name)
KeyError: '/psm_f8001f6c'
```

File: docs/stable/store/quickstart.md
Comment (SiyangShao) [inline docs/stable/store/quickstart.md:1]: The new quickstart guide provided here seems cannot work,
```
An error occurred while saving the model: [Errno 2] No such file or directory: './models/facebook/opt-1.3b/merges.txt'
[rank0]: Traceback (most recent call last):
[rank0]:   File "/home/shaosy/test/t/save.py", line 73, in download_vllm_model
[rank0]:     _run_writer(input_dir, model_name)
[rank0]:   File "/home/shaosy/test/t/save.py", line 57, in _run_writer
[rank0]:     shutil.copy(src_path, dest_path)
[rank0]:   File "/home/shaosy/test/env/lib/python3.10/shutil.py", line 417, in copy
[rank0]:     copyfile(src, dst, follow_symlinks=follow_symlinks)
[rank0]:   File "/home/shaosy/test/env/lib/python3.10/shutil.py", line 256, in copyfile
[rank0]:     with open(dst, 'wb') as fdst:
[rank0]: FileNotFoundError: [Errno 2] No such file or directory: './models/facebook/opt-1.3b/merges.txt'

[rank0]: During handling of the above exception, another exception occurred:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/home/shaosy/test/t/save.py", line 83, in <module>
[rank0]:     downloader.download_vllm_model("facebook/opt-1.3b", "float16", 1)
[rank0]:   File "/home/shaosy/test/t/save.py", line 77, in download_vllm_model
[rank0]:     shutil.rmtree(os.path.join(storage_path, model_name))
[rank0]:   File "/home/shaosy/test/env/lib/python3.10/shutil.py", line 715, in rmtree
[rank0]:     onerror(os.lstat, path, sys.exc_info())
[rank0]:   File "/home/shaosy/test/env/lib/python3.10/shutil.py", line 713, in rmtree
[rank0]:     orig_st = os.lstat(path)
[rank0]: FileNotFoundError: [Errno 2] No such file or directory: './models/facebook/opt-1.3b'
```

**Review by SiyangShao — CHANGES_REQUESTED** [review]

File: docs/stable/store/quickstart.md
Comment (future-xy) [reply docs/stable/store/quickstart.md:1]: it shloud work now

**Review by SiyangShao — APPROVED** [review]

**Pull request merged**